cargo-pgrx/src/command/upgrade.rs

```diff
@@ -0,0 +1,216 @@
+//LICENSE Portions Copyright 2019-2021 ZomboDB, LLC.
+//LICENSE
+//LICENSE Portions Copyright 2021-2023 Technology Concepts & Design, Inc.
+//LICENSE
+//LICENSE Portions Copyright 2023-2023 PgCentral Foundation, Inc. <contact@pgcentral.org>
+//LICENSE
+//LICENSE All rights reserved.
+//LICENSE
+//LICENSE Use of this source code is governed by the MIT license that can be found in the LICENSE file.
+use cargo_edit::{registry_url, Dependency};
+use eyre::eyre;
+use std::path::{Path, PathBuf};
+use toml_edit::KeyMut;
+use tracing::{debug, error, info, warn};
+
+use crate::CommandExecute;
+
+/// Upgrade pgrx crate versions in `Cargo.toml`.
+/// Defaults to latest.
+#[derive(clap::Args, Debug)]
+#[clap(author)]
+pub(crate) struct Upgrade {
+    /// Specify a version to upgrade to, rather than defaulting to the latest
+    /// version.
+    #[clap(long, short)]
+    pub(crate) target_version: Option<String>, //TODO: typed version not string
+
+    /// Path to the manifest file (usually Cargo.toml). Defaults to
+    /// "./Cargo.toml" in the working directory.
+    #[clap(long, short)]
+    pub(crate) path: Option<PathBuf>,
+
+    // Flag to permit pre-release builds
+    #[clap(long, short)]
+    pub(crate) allow_prerelease: bool,
+}
+trait DependencySource {
+    fn set_version<A: AsRef<str>>(&mut self, new_version: A);
+    fn get_version(&self) -> Option<&String>;
+}
+impl DependencySource for cargo_edit::Source {
+    fn set_version<A: AsRef<str>>(&mut self, new_version: A) {
+        match self {
+            cargo_edit::Source::Registry(reg) => reg.version = new_version.as_ref().to_string(),
+            cargo_edit::Source::Path(path) => path.version = Some(new_version.as_ref().to_string()),
+            cargo_edit::Source::Git(git) => git.version = Some(new_version.as_ref().to_string()),
```

Comment:
hmm. i am not sure the behavior for the registry and git sources is what we want. take for instance a git dependency like this:
```toml
pgrx = { git = "https://github.com/pgcentralfoundation/pgrx/" }
```
upgrade says this about it:
```
$ pgrx upgrade
INFO cargo_pgrx::command::upgrade: No version specified for pgrx, not upgrading.
```
that seems ok. but if there is a version specified, it does this:
```diff
-pgrx = { git = "https://github.com/pgcentralfoundation/pgrx/", version = "0.12.0-alpha.1" }
+pgrx = { git = "https://github.com/pgcentralfoundation/pgrx/", version = "0.11.3" }
```
and then the next time we run cargo, it fails:
```
error: failed to select a version for the requirement `pgrx = "^0.11.3"`
candidate versions found which didn't match: 0.12.0-alpha.1
```
maybe we should just error out for git or path dependencies, because we can't control their version? admittedly using a version requirement with `git` is already kinda weird and maybe the current behavior is what the user expects.

Reply:
That is a bit of an oddity, yes. I could also see it as reasonable to use the current version found in Cargo.toml at the source path - parse the manifest at the path specified, and then use the package version found in that file. 

Whereas it should probably print an error upon encountering git dependencies rather than trying to upgrade them.

Reply:
After a conversation with @thomcc  on Discord it seems safer to throw a warning upon trying to upgrade a path or git repo dependency, so that the program doesn't modify Cargo.toml under incorrect assumptions. It's probably better for software development tools to avoid trying to be too clever.

Reply:
Resolved in [2c9e997](https://github.com/pgcentralfoundation/pgrx/pull/1687/commits/2c9e997bcf521aa26c8a75d12193a70ca7e8a517)